src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -25,14 +28,20 @@ internal CompositeEndpointDataSource(IEnumerable<EndpointDataSource> dataSources
 
             _dataSources = dataSources.ToArray();
             _lock = new object();
+
+            CreateConsumerChangeToken();
         }
 
         public override IChangeToken ChangeToken
         {
             get
             {
                 EnsureInitialized();
-                return _changeToken;
+                if (_consumerChangeToken.HasChanged)
```

Comment:
This change enables to not register a callback to an inflight event which resulted in stackoverflow